Xena: Rocky 8 AIO scenario

jovial opened the pull request:
I'll rebase this when this when the test is passing. It builds on a few other open PRs.

jovial committed:
Adds Rocky 8 pulp repositories to dnf.yml

This configures the hosts to use the mirrors in the local
pulp. Prior to this change kayobe would attempt to configure
CentOS stream mirrors on a Rocky host.

jovial committed:
Rocky 8: Install a compatible version of ansible in the kolla venv

This improves support out of the box. You still need to make sure that
you install a compatible version of ansible in the kayobe virtualenv:

```
git clone https://github.com/stackhpc/stackhpc-kayobe-config -b stackhpc/wallaby
cd stackhpc-kayobe-config
pip install -U ansible-base@git+https://github.com/stackhpc/ansible@stackhpc/2.10/rocky -e .
```

The above example installs a fork of ansible with support for Rocky 8.

jovial committed:
Update etc/kayobe/kolla.yml

jovial committed:
Merge remote-tracking branch 'origin/feature/xena/rocky-dnf' into stackhpc/xena

jovial committed:
Add playbook to generate libvirt sasl secret

jovial committed:
Set compute_libvirt_sasl_password to secret from generator

jovial committed:
Follow up: Update docker baseurl to work with Rocky 8

This is the equivalent change of:

https://review.opendev.org/c/openstack/ansible-collection-kolla/+/831642

I missed that this is defined again in the ci environment. Note we
can remove this duplicate definition in Xena as the kolla globals
merging feature landed:

https://review.opendev.org/c/openstack/kayobe/+/804068

But this change is cherry-pickable to wallaby.

jovial committed:
Rocky 8: Enable libvirt on the host

jovial committed:
Make Rocky Image more usable out of the box

jovial committed:
Bump volume size to support default stackhpc LVM sizes

Should probably test with the defaults.

jovial committed:
Copy approach used for Ubuntu

https://github.com/stackhpc/stackhpc-kayobe-config/pull/138

jovial commented on etc/kayobe/overcloud-dib.yml:
```diff
@@ -71,7 +71,8 @@ overcloud_dib_host_packages_extra:
   - repo: "https://github.com/stackhpc/stackhpc-image-elements"
     local: "{{ source_checkout_path }}/stackhpc-image-elememts"
-    version: "v1.4.0"
+    # FIXME: merge and tag a new release
```
Not technically need for AIO, so could this split this out. I was just testing with the image produced by DIB. Should we switch to using the LVM based DIB images at some point?

jovial commented on terraform/aio/vm.tf:
```diff
@@ -59,35 +49,9 @@ resource "openstack_compute_instance_v2" "kayobe-aio" {
     uuid                  = data.openstack_images_image_v2.image.id
     source_type           = "image"
-    volume_size           = 100
+    volume_size           = 200
```
This was for the stackhpc LVM layout with the DIB image

jovial commented:
Is it best to land this change on wallaby? Seems easier to merge up than down? I assumed I should target the latest release, but maybe this was a mistake?

markgoddard commented:
> Is it best to land this change on wallaby? Seems easier to merge up than down? I assumed I should target the latest release, but maybe this was a mistake?

It is possible to merge up, but not down. You can cherry pick down if necessary though. It might be safer to start with Xena since it's not in use yet.

Will we have all the necessary dependencies in Wallaby?

markgoddard commented on etc/kayobe/ansible/templates/stackhpc-secrets.yml.j2:
```diff
@@ -0,0 +1,3 @@
+---
+# SASL password when using libvirt on the host
+secrets_libvirt_sasl_password: "{{ secrets_libvirt_sasl_password | default(lookup('password', '/dev/null'), true) }}"
```
Could you not use the password lookup plugin, and write the result to a file?

jovial replied:
Could work. Are you imagining to just use that for the AIO? Or would we write it some location in kayobe config? Something like `{{ kayobe_config_env_path }}/secrets/libvirt_sasl_password`?

markgoddard commented on etc/kayobe/compute.yml:
```diff
@@ -179,6 +179,15 @@
 # compute_qemu_conf_extra.
 #compute_qemu_conf:
 
+# Whether to enable libvirt SASL authentication. Default is true.
```
Do we need a config sync somewhere?

jovial replied:
Think so, will check when these keys landed upstream.

jovial replied:
Trying to fix upstream here: https://review.opendev.org/c/openstack/kayobe/+/856454. Will also need syncing to kayobe-config.

markgoddard commented on etc/kayobe/compute.yml:
```diff
@@ -179,6 +179,15 @@
+# NOTE: You can use:
+#   kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/generate-stackhpc-secrets.yml
+# to generate secrets_libvirt_sasl_password.
+compute_libvirt_sasl_password: "{{ secrets_libvirt_sasl_password }}"
```
Probably needs to be in the ci-aio environment?

jovial replied:
Can we keep it in the base config? It isn't used unless you set kolla_enable_nova_libvirt_container to false.

markgoddard commented on etc/kayobe/kolla.yml:
```diff
@@ -506,6 +506,7 @@ kolla_build_args:
 #kolla_enable_kibana:
 #kolla_enable_kuryr:
 #kolla_enable_loadbalancer:
+kolla_enable_nova_libvirt_container: "{{ not os_distribution in ['rocky'] }}"
```
X not in Y?

jovial replied:
Much nicer, thanks.

markgoddard commented on terraform/aio/templates/userdata.cfg.tpl:
```diff
@@ -2,3 +2,11 @@
+  - 'echo "10.205.3.187 pulp-server pulp-server.internal.sms-cloud" >> /etc/hosts'
+# Configure SSH keys here, to avoid creating an ephemeral keypair.
+# This means only the instance needs to be cleaned up if the destroy fails.
+ssh_authorized_keys:
```
IIRC this didn't work for me in the Ubuntu job, unsure which user got the key.

jovial replied:
Cheers, I didn't check this as I only saw the ubuntu PR late and thought I'd adopt the method from there.

jovial replied:
Fixed with https://github.com/stackhpc/stackhpc-kayobe-config/pull/155/commits/ff692fa94a46d675a7ffc114d9c464c914192db7

jovial committed:
Use templatefile on userdata

This is so we can inject the public key within the userdata.
Previously we were using the file function which does not
do any templating. This meant that we had the literal string
"file(var.ssh_public_key)" in our userdata and consequently
could not login.

jovial committed:
Remove commented out config

jovial committed:
Fix warning about undeclared variables

jovial committed:
Address code review comment

jovial committed:
Switch to using password lookup for sasl password

jovial committed:
Remove generate secrets step

jovial committed:
Use facts to pick dnf repositories

jovial commented on .github/workflows/stackhpc-pull-request.yml:
```diff
@@ -57,4 +57,34 @@ jobs:
+      os_distribution: rocky
+      neutron_plugin: ovs
+      # FIXME: This image is ancient, but is the only Rocky image uploaded?
+      vm_image: Rocky-8-GenericCloud-8.5-20211114.2.x86_64
```
```suggestion
      vm_image: Rocky8
```

jovial committed:
Try another image

jovial committed:
Try using cloud-user

jovial committed:
Update CI repository overrides for rocky

jovial committed:
Remove hardcoding of bootstrap user

jovial committed:
Merge remote-tracking branch 'origin/stackhpc/xena' into HEAD

jovial committed:
Merge branch 'stackhpc/xena' into feature/xena/aio-rocky8

jovial committed:
Fix bad reference to github inputs

jovial commented on .github/workflows/stackhpc-all-in-one.yml:
```diff
@@ -56,6 +64,12 @@ jobs:
+      - name: Set bootstrap user output
+        id: bootstrap_user
+        run: |
+          user=${{ github.event.inputs.vm_user }}
```
```suggestion
          user=${{ inputs.vm_user }}
```

jovial committed:
Fix bootstrap user set to empty string

jovial commented on .github/workflows/stackhpc-all-in-one.yml:
```diff
@@ -151,10 +163,13 @@ jobs:
           ---
+          os_distribution: ${{ env.OS_DISTRIBUTION }}
           kolla_enable_ovn: ${{ env.ENABLE_OVN }}
+          bootstrap_user: ${{ steps.name.ouputs.value }}
```
```suggestion
          bootstrap_user: ${{ steps.bootstrap_user.ouputs.value }}
```

jovial committed:
Fix bootstrap user set to empty string: take 2

jovial committed:
Fix empty bootstrap user: take 3

jovial commented on etc/kayobe/stackhpc.yml:
```diff
@@ -103,6 +103,18 @@ stackhpc_repo_rabbitmq_server_version: "{{ stackhpc_repo_distribution }}"
 stackhpc_repo_treasuredata_4_version: "{{ stackhpc_repo_distribution }}"
 
+# Rocky 8 BaseOS
+stackhpc_repo_rocky_baseos_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8/BaseOS/x86_64/os/{{ stackhpc_repo_rocky_baseos_version }}"
```
```suggestion
stackhpc_repo_rocky_baseos_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8.6/BaseOS/x86_64/os/{{ stackhpc_repo_rocky_baseos_version }}"
```

jovial commented on etc/kayobe/stackhpc.yml:
```diff
@@ -103,6 +103,18 @@ stackhpc_repo_rabbitmq_server_version: "{{ stackhpc_repo_distribution }}"
+stackhpc_repo_rocky_baseos_version: "{{ stackhpc_repo_distribution }}"
+
+# Rocky 8 AppStream
+stackhpc_repo_rocky_appstream_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8/AppStream/x86_64/os/{{ stackhpc_repo_rocky_appstream_version }}"
```
```suggestion
stackhpc_repo_rocky_appstream_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8.6/AppStream/x86_64/os/{{ stackhpc_repo_rocky_appstream_version }}"
```

jovial commented on etc/kayobe/stackhpc.yml:
```diff
@@ -103,6 +103,18 @@ stackhpc_repo_rabbitmq_server_version: "{{ stackhpc_repo_distribution }}"
+stackhpc_repo_rocky_appstream_version: "{{ stackhpc_repo_distribution }}"
+
+# Rocky 8 extras
+stackhpc_repo_rocky_extras_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8/extras/x86_64/os/{{ stackhpc_repo_rocky_extras_version }}"
```
```suggestion
stackhpc_repo_rocky_extras_url: "{{ stackhpc_repo_mirror_url }}/pulp/content/rocky/8.6/extras/x86_64/os/{{ stackhpc_repo_rocky_extras_version }}"
```

jovial committed:
Fix internal pulp URL

Internal pulp URLs include minor version

jovial committed:
Workaround network interfaces restart failing

jovial committed:
Revert to using terraform to run workaround

jovial committed:
Add missing variable

jovial committed:
Restore configure local networking

jovial committed:
A few fixes for the Rocky 8 image

jovial committed:
Fixups

jovial committed:
Re-enable other jobs to check I havent broken anything

jovial committed:
Merge remote-tracking branch 'origin/stackhpc/xena' into HEAD

jovial committed:
Merge branch 'stackhpc/xena' into feature/xena/aio-rocky8

jovial committed:
Remove workaround for old rocky8 image

jovial commented:
Closing in favour of https://github.com/stackhpc/stackhpc-kayobe-config/pull/235.
